File: sdks/java/src/main/java/to/renamed/sdk/RenamedClient.java
package to.renamed.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Client for the renamed.to API.
 *
 * <p>The RenamedClient provides access to all renamed.to API operations including
 * file renaming, PDF splitting, and data extraction.</p>
 *
 * <p>Basic usage:</p>
 * <pre>{@code
 * RenamedClient client = new RenamedClient("rt_your_api_key");
 *
 * // Rename a file
 * RenameResult result = client.rename(Path.of("invoice.pdf"), null);
 * System.out.println("Suggested: " + result.getSuggestedFilename());
 *
 * // Get user info
 * User user = client.getUser();
 * System.out.println("Credits: " + user.getCredits());
 * }</pre>
 *
 * <p>With custom configuration:</p>
 * <pre>{@code
 * RenamedClient client = RenamedClient.builder("rt_your_api_key")
 *     .baseUrl("https://custom-api.example.com")
 *     .timeout(Duration.ofSeconds(60))
 *     .maxRetries(3)
 *     .build();
 * }</pre>
 */
public class RenamedClient {

    private static final String DEFAULT_BASE_URL = "https://www.renamed.to/api/v1";
    private static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(30);
    private static final int DEFAULT_MAX_RETRIES = 2;
    private static final String CRLF = "\r\n";

    private final String apiKey;
    private final String baseUrl;
    private final Duration timeout;
    private final int maxRetries;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final boolean debug;
    private final Logger logger;

    /**
     * Creates a new RenamedClient with the specified API key.
     *
     * @param apiKey your renamed.to API key (starts with "rt_")
     */
    public RenamedClient(String apiKey) {
        this(apiKey, DEFAULT_BASE_URL, DEFAULT_TIMEOUT, DEFAULT_MAX_RETRIES, null, false, null);
    }

    /**
     * Creates a new RenamedClient with full configuration.
     */
    private RenamedClient(String apiKey, String baseUrl, Duration timeout, int maxRetries,
                          HttpClient httpClient, boolean debug, Logger logger) {
        if (apiKey == null || apiKey.isEmpty()) {
            throw new IllegalArgumentException("API key is required");
        }
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.timeout = timeout;
        this.maxRetries = maxRetries;
        this.httpClient = httpClient != null ? httpClient : HttpClient.newBuilder()
                .connectTimeout(timeout)
                .build();
        this.objectMapper = new ObjectMapper();
        this.debug = debug;
        this.logger = debug ? (logger != null ? logger : new DebugLogger()) : null;

        if (this.debug && this.logger != null) {
            this.logger.log("Initialized with API key " + DebugLogger.maskApiKey(apiKey));
        }
    }

    /**
     * Creates a new builder for configuring a RenamedClient.
     *
     * @param apiKey your renamed.to API key
     * @return a new builder instance
     */
    public static Builder builder(String apiKey) {
        return new Builder(apiKey);
    }

    /**
     * Returns the current user profile and credits.
     *
     * <p>Example:</p>
     * <pre>{@code
     * User user = client.getUser();
     * System.out.println("Email: " + user.getEmail());
     * System.out.println("Credits: " + user.getCredits());
     * }</pre>
     *
     * @return the user profile
     * @throws AuthenticationError if the API key is invalid
     * @throws RenamedError if the request fails
     */
    public User getUser() {
        String response = request("GET", "/user", null, null);
        return parseResponse(response, User.class);
    }

    /**
     * Renames a file using AI.
     *
     * <p>Example:</p>
     * <pre>{@code
     * RenameResult result = client.rename(Path.of("document.pdf"), null);
     * System.out.println("Original: " + result.getOriginalFilename());
     * System.out.println("Suggested: " + result.getSuggestedFilename());
     * }</pre>
     *
     * @param file the file to rename
     * @param options optional rename settings
     * @return the rename result with suggested filename
     * @throws ValidationError if the file format is not supported
     * @throws InsufficientCreditsError if the account has no credits
     * @throws RenamedError if the request fails
     */
    public RenameResult rename(Path file, RenameOptions options) {
        Map<String, String> fields = new HashMap<>();
        if (options != null && options.getTemplate() != null) {
            fields.put("template", options.getTemplate());
        }

        String response = uploadFile("/rename", file, fields);
        return parseResponse(response, RenameResult.class);
    }

    /**
     * Renames a file from a byte array.
     *
     * @param content the file content
     * @param filename the filename
     * @param options optional rename settings
     * @return the rename result with suggested filename
     */
    public RenameResult rename(byte[] content, String filename, RenameOptions options) {
        Map<String, String> fields = new HashMap<>();
        if (options != null && options.getTemplate() != null) {
            fields.put("template", options.getTemplate());
        }

        String response = uploadFile("/rename", content, filename, fields);
        return parseResponse(response, RenameResult.class);
    }

    /**
     * Renames a file from an InputStream.
     *
     * @param inputStream the input stream containing file content
     * @param filename the filename
     * @param options optional rename settings
     * @return the rename result with suggested filename
     * @throws RenamedError if the request fails
     */
    public RenameResult rename(InputStream inputStream, String filename, RenameOptions options) {
        try {
            byte[] content = inputStream.readAllBytes();
            return rename(content, filename, options);
        } catch (IOException e) {
            throw new NetworkError("Failed to read input stream: " + e.getMessage(), e);
        }
    }

    /**
     * Splits a PDF into multiple documents.
     *
     * <p>This operation is asynchronous. The returned {@link AsyncJob} can be
     * used to poll for completion and retrieve results.</p>
     *
     * <p>Example:</p>
     * <pre>{@code
     * PdfSplitOptions options = new PdfSplitOptions()
     *     .withMode(SplitMode.AUTO);
     *
     * AsyncJob job = client.pdfSplit(Path.of("multi-page.pdf"), options);
     *
     * // Wait for completion with progress updates
     * PdfSplitResult result = job.await(status -> {
     *     System.out.println("Progress: " + status.getProgress() + "%");
     * });
     *
     * for (SplitDocument doc : result.getDocuments()) {
     *     System.out.println(doc.getFilename());
     * }
     * }</pre>
     *
     * @param file the PDF file to split
     * @param options optional split settings
     * @return an async job that can be polled for results
     * @throws ValidationError if the file is not a valid PDF
     * @throws InsufficientCreditsError if the account has no credits
     * @throws RenamedError if the request fails
     */
    public AsyncJob pdfSplit(Path file, PdfSplitOptions options) {
        Map<String, String> fields = new HashMap<>();
        if (options != null) {
            if (options.getMode() != null) {
                fields.put("mode", options.getMode().getValue());
            }
            if (options.getPagesPerSplit() > 0) {
                fields.put("pagesPerSplit", String.valueOf(options.getPagesPerSplit()));
            }
        }

        String response = uploadFile("/pdf-split", file, fields);
        return createAsyncJob(response);
    }

    /**
     * Splits a PDF from a byte array.
     *
     * @param content the PDF file content
     * @param filename the filename
     * @param options optional split settings
     * @return an async job that can be polled for results
     */
    public AsyncJob pdfSplit(byte[] content, String filename, PdfSplitOptions options) {
        Map<String, String> fields = new HashMap<>();
        if (options != null) {
            if (options.getMode() != null) {
                fields.put("mode", options.getMode().getValue());
            }
            if (options.getPagesPerSplit() > 0) {
                fields.put("pagesPerSplit", String.valueOf(options.getPagesPerSplit()));
            }
        }

        String response = uploadFile("/pdf-split", content, filename, fields);
        return createAsyncJob(response);
    }

    /**
     * Splits a PDF from an InputStream.
     *
     * @param inputStream the input stream containing PDF content
     * @param filename the filename
     * @param options optional split settings
     * @return an async job that can be polled for results
     */
    public AsyncJob pdfSplit(InputStream inputStream, String filename, PdfSplitOptions options) {
        try {
            byte[] content = inputStream.readAllBytes();
            return pdfSplit(content, filename, options);
        } catch (IOException e) {
            throw new NetworkError("Failed to read input stream: " + e.getMessage(), e);
        }
    }

    /**
     * Extracts structured data from a document.
     *
     * <p>Example:</p>
     * <pre>{@code
     * ExtractOptions options = new ExtractOptions()
     *     .withPrompt("Extract the invoice number, date, and total amount");
     *
     * ExtractResult result = client.extract(Path.of("invoice.pdf"), options);
     *
     * Map<String, Object> data = result.getData();
     * System.out.println("Invoice: " + data.get("invoiceNumber"));
     * System.out.println("Total: $" + data.get("total"));
     * }</pre>
     *
     * @param file the document to extract data from
     * @param options extraction options (prompt or schema)
     * @return the extraction result with structured data
     * @throws ValidationError if the file format is not supported
     * @throws InsufficientCreditsError if the account has no credits
     * @throws RenamedError if the request fails
     */
    public ExtractResult extract(Path file, ExtractOptions options) {
        Map<String, String> fields = new HashMap<>();
        if (options != null) {
            if (options.getPrompt() != null) {
                fields.put("prompt", options.getPrompt());
            }
            if (options.getSchema() != null) {
                try {
                    fields.put("schema", objectMapper.writeValueAsString(options.getSchema()));
                } catch (IOException e) {
                    throw new ValidationError("Failed to serialize schema: " + e.getMessage());
                }
            }
        }

        String response = uploadFile("/extract", file, fields);
        return parseResponse(response, ExtractResult.class);
    }

    /**
     * Extracts structured data from a byte array.
     *
     * @param content the file content
     * @param filename the filename
     * @param options extraction options
     * @return the extraction result with structured data
     */
    public ExtractResult extract(byte[] content, String filename, ExtractOptions options) {
        Map<String, String> fields = new HashMap<>();
        if (options != null) {
            if (options.getPrompt() != null) {
                fields.put("prompt", options.getPrompt());
            }
            if (options.getSchema() != null) {
                try {
                    fields.put("schema", objectMapper.writeValueAsString(options.getSchema()));
                } catch (IOException e) {
                    throw new ValidationError("Failed to serialize schema: " + e.getMessage());
                }
            }
        }

        String response = uploadFile("/extract", content, filename, fields);
        return parseResponse(response, ExtractResult.class);
    }

    /**
     * Extracts structured data from an InputStream.
     *
     * @param inputStream the input stream containing file content
     * @param filename the filename
     * @param options extraction options
     * @return the extraction result with structured data
     */
    public ExtractResult extract(InputStream inputStream, String filename, ExtractOptions options) {
        try {
            byte[] content = inputStream.readAllBytes();
            return extract(content, filename, options);
        } catch (IOException e) {
            throw new NetworkError("Failed to read input stream: " + e.getMessage(), e);
        }
    }

    /**
     * Downloads a file from a URL (e.g., split document).
     *
     * <p>Example:</p>
     * <pre>{@code
     * PdfSplitResult result = job.await();
     * for (SplitDocument doc : result.getDocuments()) {
     *     byte[] content = client.downloadFile(doc.getDownloadUrl());
     *     Files.write(Path.of(doc.getFilename()), content);
     * }
     * }</pre>
     *
     * @param url the download URL
     * @return the file content as a byte array
     * @throws RenamedError if the download fails
     */
    public byte[] downloadFile(String url) {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .header("Authorization", "Bearer " + apiKey)
                    .GET()
                    .timeout(timeout)
                    .build();

            long startTime = System.currentTimeMillis();
            HttpResponse<byte[]> response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray());
            long elapsed = System.currentTimeMillis() - startTime;

            // Extract path from URL for logging (avoid logging full URL which may contain tokens)
            String path = URI.create(url).getPath();
            if (logger != null) {
                logger.log("GET " + path + " -> " + response.statusCode() + " (" + elapsed + "ms)");
            }

            if (response.statusCode() >= 400) {
                throw RenamedError.fromHttpStatus(response.statusCode(),
                        "HTTP " + response.statusCode(), null);
            }

            if (logger != null) {
                logger.log("Download: " + DebugLogger.formatSize(response.body().length));
            }

            return response.body();
        } catch (IOException e) {
            throw new NetworkError("Download failed: " + e.getMessage(), e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new NetworkError("Download interrupted", e);
        }
    }

    private String buildUrl(String path) {
        if (path.startsWith("http://") || path.startsWith("https://")) {
            return path;
        }
        if (path.startsWith("/")) {
            return baseUrl + path;
        }
        return baseUrl + "/" + path;
    }

    private String request(String method, String path, byte[] body, String contentType) {
        String url = buildUrl(path);
        Exception lastException = null;

        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            try {
                HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                        .uri(URI.create(url))
                        .header("Authorization", "Bearer " + apiKey)
                        .timeout(timeout);

                if (body != null) {
                    requestBuilder.method(method, HttpRequest.BodyPublishers.ofByteArray(body));
                    if (contentType != null) {
                        requestBuilder.header("Content-Type", contentType);
                    }
                } else {
                    requestBuilder.method(method, HttpRequest.BodyPublishers.noBody());
                }

                long startTime = System.currentTimeMillis();
                HttpResponse<String> response = httpClient.send(requestBuilder.build(),
                        HttpResponse.BodyHandlers.ofString());
                long elapsed = System.currentTimeMillis() - startTime;

                if (logger != null) {
                    logger.log(method + " " + path + " -> " + response.statusCode() + " (" + elapsed + "ms)");
                }

                if (response.statusCode() >= 400) {
                    Map<String, Object> payload = parseErrorPayload(response.body());
                    throw RenamedError.fromHttpStatus(response.statusCode(),
                            "HTTP " + response.statusCode(), payload);
                }

                return response.body();
            } catch (IOException e) {
                lastException = e;
                if (attempt < maxRetries) {
                    long backoffMs = (1L << attempt) * 100;
                    if (logger != null) {
                        logger.log("Retry attempt " + (attempt + 1) + "/" + maxRetries + ", waiting " + backoffMs + "ms");
                    }
                    try {
                        Thread.sleep(backoffMs);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new NetworkError("Request interrupted", ie);
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new NetworkError("Request interrupted", e);
            }
        }

        throw new NetworkError("Request failed after " + (maxRetries + 1) + " attempts: " +
                (lastException != null ? lastException.getMessage() : "unknown error"), lastException);
    }

    private String uploadFile(String path, Path file, Map<String, String> fields) {
        try {
            byte[] content = Files.readAllBytes(file);
            String filename = file.getFileName().toString();
            return uploadFile(path, content, filename, fields);
        } catch (IOException e) {
            throw new NetworkError("Failed to read file: " + e.getMessage(), e);
        }
    }

    private String uploadFile(String path, byte[] content, String filename, Map<String, String> fields) {
        if (logger != null) {
            logger.log("Upload: " + filename + " (" + DebugLogger.formatSize(content.length) + ")");
        }
        String boundary = "----RenamedBoundary" + UUID.randomUUID().toString().replace("-", "");
        byte[] body = buildMultipartBody(boundary, content, filename, fields);
        return request("POST", path, body, "multipart/form-data; boundary=" + boundary);
    }

    private byte[] buildMultipartBody(String boundary, byte[] fileContent, String filename, Map<String, String> fields) {
        try {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            PrintWriter writer = new PrintWriter(new OutputStreamWriter(baos, StandardCharsets.UTF_8), true);

            // Add file field
            writer.append("--").append(boundary).append(CRLF);
            writer.append("Content-Disposition: form-data; name=\"file\"; filename=\"")
                    .append(filename).append("\"").append(CRLF);
            writer.append("Content-Type: ").append(getMimeType(filename)).append(CRLF);
            writer.append(CRLF).flush();
            baos.write(fileContent);
            baos.flush();
            writer.append(CRLF);

            // Add additional fields
            for (Map.Entry<String, String> entry : fields.entrySet()) {
                writer.append("--").append(boundary).append(CRLF);
                writer.append("Content-Disposition: form-data; name=\"")
                        .append(entry.getKey()).append("\"").append(CRLF);
                writer.append(CRLF);
                writer.append(entry.getValue()).append(CRLF);
            }

            // End boundary
            writer.append("--").append(boundary).append("--").append(CRLF);
            writer.flush();

            return baos.toByteArray();
        } catch (IOException e) {
            throw new NetworkError("Failed to build multipart body: " + e.getMessage(), e);
        }
    }

    private String getMimeType(String filename) {
        String lower = filename.toLowerCase();
        if (lower.endsWith(".pdf")) {
            return "application/pdf";
        }
        if (lower.endsWith(".jpg") || lower.endsWith(".jpeg")) {
            return "image/jpeg";
        }
        if (lower.endsWith(".png")) {
            return "image/png";
        }
        if (lower.endsWith(".tiff") || lower.endsWith(".tif")) {
            return "image/tiff";
        }
        return "application/octet-stream";
    }

    private <T> T parseResponse(String body, Class<T> type) {
        try {
            return objectMapper.readValue(body, type);
        } catch (IOException e) {
            throw new RenamedError("Failed to parse response: " + e.getMessage(), "PARSE_ERROR");
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseErrorPayload(String body) {
        try {
            return objectMapper.readValue(body, Map.class);
        } catch (IOException e) {
            return null;
        }
    }

    private AsyncJob createAsyncJob(String response) {
        try {
            JsonNode node = objectMapper.readTree(response);
            String statusUrl = node.get("statusUrl").asText();
            return new AsyncJob(httpClient, objectMapper, apiKey, statusUrl, logger);
        } catch (IOException e) {
            throw new RenamedError("Failed to parse async job response: " + e.getMessage(), "PARSE_ERROR");
        }
    }

    /**
     * Builder for configuring a RenamedClient.
     */
    public static class Builder {
        private final String apiKey;
        private String baseUrl = DEFAULT_BASE_URL;
        private Duration timeout = DEFAULT_TIMEOUT;
        private int maxRetries = DEFAULT_MAX_RETRIES;
        private HttpClient httpClient;
        private boolean debug = false;
        private Logger logger;

        private Builder(String apiKey) {
            this.apiKey = apiKey;
        }

        /**
         * Sets a custom base URL.
         *
         * @param baseUrl the API base URL
         * @return this builder
         */
        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        /**
         * Sets the request timeout.
         *
         * @param timeout the timeout duration
         * @return this builder
         */
        public Builder timeout(Duration timeout) {
            this.timeout = timeout;
            return this;
        }

        /**
         * Sets the maximum number of retries for failed requests.
         *
         * @param maxRetries the maximum retry count
         * @return this builder
         */
        public Builder maxRetries(int maxRetries) {
            this.maxRetries = maxRetries;
            return this;
        }

        /**
         * Sets a custom HTTP client.
         *
         * @param httpClient the HTTP client to use
         * @return this builder
         */
        public Builder httpClient(HttpClient httpClient) {
            this.httpClient = httpClient;
            return this;
        }

        /**
         * Enables or disables debug logging.
         *
         * <p>When enabled, the SDK logs HTTP requests, responses, retries,
         * file uploads, and async job polling. API keys are masked in logs.</p>
         *
         * <p>Example:</p>
         * <pre>{@code
         * RenamedClient client = RenamedClient.builder("rt_your_api_key")
         *     .debug(true)
         *     .build();
         * }</pre>
         *
         * @param debug true to enable debug logging
         * @return this builder
         */
        public Builder debug(boolean debug) {
            this.debug = debug;
            return this;
        }

        /**
         * Sets a custom logger for debug output.
         *
         * <p>If not set and debug is enabled, logs are written to System.out.
         * The logger can integrate with any logging framework.</p>
         *
         * <p>Example with SLF4J:</p>
         * <pre>{@code
         * import org.slf4j.LoggerFactory;
         *
         * RenamedClient client = RenamedClient.builder("rt_your_api_key")
         *     .debug(true)
         *     .logger(LoggerFactory.getLogger(RenamedClient.class)::info)
         *     .build();
         * }</pre>
         *
         * @param logger the logger to use for debug output
         * @return this builder
         */
        public Builder logger(Logger logger) {
            this.logger = logger;
            return this;
        }

        /**
         * Builds the RenamedClient with the configured settings.
         *
         * @return the configured client
         */
        public RenamedClient build() {
            return new RenamedClient(apiKey, baseUrl, timeout, maxRetries, httpClient, debug, logger);
        }
    }
}

File: sdks/swift/Sources/Renamed/RenamedClient.swift
import Foundation

/// Configuration options for the Renamed client
public struct RenamedClientOptions: Sendable {
    /// API key for authentication (starts with rt_)
    public let apiKey: String

    /// Base URL for the API
    public let baseUrl: String

    /// Request timeout in seconds
    public let timeout: TimeInterval

    /// Maximum number of retries for failed requests
    public let maxRetries: Int

    /// Enable debug logging (uses print() by default)
    public let debug: Bool

    /// Custom logger implementation (overrides debug flag when set)
    public let logger: (any RenamedLogger)?

    /// Default base URL
    public static let defaultBaseUrl = "https://www.renamed.to/api/v1"

    /// Default timeout in seconds
    public static let defaultTimeout: TimeInterval = 30

    /// Default max retries
    public static let defaultMaxRetries = 2

    public init(
        apiKey: String,
        baseUrl: String = RenamedClientOptions.defaultBaseUrl,
        timeout: TimeInterval = RenamedClientOptions.defaultTimeout,
        maxRetries: Int = RenamedClientOptions.defaultMaxRetries,
        debug: Bool = false,
        logger: (any RenamedLogger)? = nil
    ) {
        self.apiKey = apiKey
        self.baseUrl = baseUrl
        self.timeout = timeout
        self.maxRetries = maxRetries
        self.debug = debug
        self.logger = logger
    }
}

/// Main client for the renamed.to API
public final class RenamedClient: @unchecked Sendable {
    private let apiKey: String
    private let baseUrl: String
    private let timeout: TimeInterval
    private let maxRetries: Int
    private let session: URLSession
    private let logger: (any RenamedLogger)?

    /// Create a new client with the given options
    public init(options: RenamedClientOptions) throws {
        guard !options.apiKey.isEmpty else {
            throw RenamedError.authentication(message: "API key is required")
        }

        self.apiKey = options.apiKey
        self.baseUrl = options.baseUrl
        self.timeout = options.timeout
        self.maxRetries = options.maxRetries

        // Set up logger: custom logger takes precedence, then debug flag
        if let customLogger = options.logger {
            self.logger = customLogger
        } else if options.debug {
            self.logger = DefaultLogger.shared
        } else {
            self.logger = nil
        }

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout * 2
        self.session = URLSession(configuration: config)
    }

    /// Create a new client with just an API key using default options
    public convenience init(apiKey: String) throws {
        try self.init(options: RenamedClientOptions(apiKey: apiKey))
    }

    // MARK: - Logging Helpers

    /// Log a debug message if logging is enabled
    internal func log(_ message: String) {
        logger?.debug("\(LogHelper.prefix) \(message)")
    }

    // MARK: - Public API Methods

    /// Get current user profile and credits
    ///
    /// ```swift
    /// let user = try await client.getUser()
    /// print("Credits remaining: \(user.credits ?? 0)")
    /// ```
    public func getUser() async throws -> User {
        try await request(path: "/user", method: "GET")
    }

    /// Rename a file using AI
    ///
    /// ```swift
    /// let file = FileInput(url: URL(fileURLWithPath: "invoice.pdf"))
    /// let result = try await client.rename(file: file)
    /// print(result.suggestedFilename) // "2025-01-15_AcmeCorp_INV-12345.pdf"
    /// ```
    ///
    /// - Parameters:
    ///   - file: The file to analyze
    ///   - options: Optional rename options (template, etc.)
    /// - Returns: Rename result with suggested filename
    public func rename(
        file: FileInput,
        options: RenameOptions? = nil
    ) async throws -> RenameResult {
        var additionalFields: [String: String] = [:]

        if let template = options?.template {
            additionalFields["template"] = template
        }

        return try await uploadFile(
            path: "/rename",
            file: file,
            additionalFields: additionalFields.isEmpty ? nil : additionalFields
        )
    }

    /// Split a PDF into multiple documents
    ///
    /// Returns an AsyncJob that can be polled for completion.
    ///
    /// ```swift
    /// let file = try FileInput(url: URL(fileURLWithPath: "multi-page.pdf"))
    /// let job = try await client.pdfSplit(file: file, options: PdfSplitOptions(mode: .auto))
    /// let result = try await job.wait { status in
    ///     print("Progress: \(status.progress ?? 0)%")
    /// }
    /// for doc in result.documents {
    ///     print("\(doc.filename): \(doc.downloadUrl)")
    /// }
    /// ```
    ///
    /// - Parameters:
    ///   - file: The PDF file to split
    ///   - options: Split options (mode, pages per split)
    /// - Returns: An AsyncJob that resolves to PdfSplitResult
    public func pdfSplit(
        file: FileInput,
        options: PdfSplitOptions? = nil
    ) async throws -> AsyncJob<PdfSplitResult> {
        var additionalFields: [String: String] = [:]

        if let mode = options?.mode {
            additionalFields["mode"] = mode.rawValue
        }
        if let pagesPerSplit = options?.pagesPerSplit {
            additionalFields["pagesPerSplit"] = String(pagesPerSplit)
        }

        struct JobResponse: Decodable {
            let statusUrl: String
        }

        let response: JobResponse = try await uploadFile(
            path: "/pdf-split",
            file: file,
            additionalFields: additionalFields.isEmpty ? nil : additionalFields
        )

        return AsyncJob<PdfSplitResult>(client: self, statusUrl: response.statusUrl)
    }

    /// Extract structured data from a document
    ///
    /// ```swift
    /// let file = try FileInput(url: URL(fileURLWithPath: "invoice.pdf"))
    /// let result = try await client.extract(
    ///     file: file,
    ///     options: ExtractOptions(prompt: "Extract invoice number, date, and total amount")
    /// )
    /// print(result.data)
    /// ```
    ///
    /// - Parameters:
    ///   - file: The document to extract from
    ///   - options: Extract options (schema, prompt)
    /// - Returns: Extracted data with confidence score
    public func extract(
        file: FileInput,
        options: ExtractOptions? = nil
    ) async throws -> ExtractResult {
        var additionalFields: [String: String] = [:]

        if let schema = options?.schema {
            let data = try JSONSerialization.data(withJSONObject: schema)
            additionalFields["schema"] = String(data: data, encoding: .utf8)
        }
        if let prompt = options?.prompt {
            additionalFields["prompt"] = prompt
        }

        return try await uploadFile(
            path: "/extract",
            file: file,
            additionalFields: additionalFields.isEmpty ? nil : additionalFields
        )
    }

    /// Download a file from a URL (e.g., split document)
    ///
    /// ```swift
    /// let result = try await job.wait()
    /// for doc in result.documents {
    ///     let data = try await client.downloadFile(url: doc.downloadUrl)
    ///     try data.write(to: URL(fileURLWithPath: doc.filename))
    /// }
    /// ```
    ///
    /// - Parameter url: The download URL
    /// - Returns: File data
    public func downloadFile(url: String) async throws -> Data {
        var request = URLRequest(url: URL(string: url)!)
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw RenamedError.network(message: "Invalid response")
        }

        guard httpResponse.statusCode == 200 else {
            throw RenamedError.fromHTTPStatus(
                httpResponse.statusCode,
                message: HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode)
            )
        }

        return data
    }

    // MARK: - Internal Request Methods

    /// Make an authenticated JSON request to the API
    func request<T: Decodable>(
        path: String,
        method: String,
        body: Data? = nil
    ) async throws -> T {
        let url = buildURL(path: path)
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

        if let body = body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = body
        }

        // Extract path for logging (strip base URL if present)
        let logPath = extractPath(from: path)

        return try await executeWithRetry(request: request, method: method, path: logPath)
    }

    /// Upload a file with multipart form data
    private func uploadFile<T: Decodable>(
        path: String,
        file: FileInput,
        fieldName: String = "file",
        additionalFields: [String: String]? = nil
    ) async throws -> T {
        // Log file upload details
        log("Upload: \(file.filename) (\(LogHelper.formatFileSize(file.data.count)))")

        let url = buildURL(path: path)
        let boundary = UUID().uuidString

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue(
            "multipart/form-data; boundary=\(boundary)",
            forHTTPHeaderField: "Content-Type"
        )

        var body = Data()

        // Add file field
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append(
            "Content-Disposition: form-data; name=\"\(fieldName)\"; filename=\"\(file.filename)\"\r\n"
                .data(using: .utf8)!
        )
        body.append("Content-Type: \(file.mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(file.data)
        body.append("\r\n".data(using: .utf8)!)

        // Add additional fields
        if let fields = additionalFields {
            for (key, value) in fields {
                body.append("--\(boundary)\r\n".data(using: .utf8)!)
                body.append(
                    "Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n".data(using: .utf8)!
                )
                body.append("\(value)\r\n".data(using: .utf8)!)
            }
        }

        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        // Extract path for logging
        let logPath = extractPath(from: path)

        return try await executeWithRetry(request: request, method: "POST", path: logPath)
    }

    // MARK: - Private Helpers

    private func buildURL(path: String) -> URL {
        if path.hasPrefix("http://") || path.hasPrefix("https://") {
            return URL(string: path)!
        }

        let normalizedPath = path.hasPrefix("/") ? path : "/\(path)"
        return URL(string: "\(baseUrl)\(normalizedPath)")!
    }

    private func executeWithRetry<T: Decodable>(
        request: URLRequest,
        method: String = "GET",
        path: String = ""
    ) async throws -> T {
        var lastError: Error?
        var attempts = 0

        while attempts <= maxRetries {
            let startTime = CFAbsoluteTimeGetCurrent()

            do {
                let (data, response) = try await session.data(for: request)

                guard let httpResponse = response as? HTTPURLResponse else {
                    throw RenamedError.network(message: "Invalid response")
                }

                // Log the request and response
                log("\(method) \(path) -> \(httpResponse.statusCode) (\(LogHelper.formatDuration(startTime)))")

                if httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 {
                    let decoder = JSONDecoder()
                    return try decoder.decode(T.self, from: data)
                }

                // Parse error response
                let errorMessage = parseErrorMessage(from: data) ??
                    HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode)
                let payload = try? JSONSerialization.jsonObject(with: data) as? [String: Any]

                let error = RenamedError.fromHTTPStatus(
                    httpResponse.statusCode,
                    message: errorMessage,
                    payload: payload
                )

                // Don't retry 4xx errors
                if httpResponse.statusCode >= 400 && httpResponse.statusCode < 500 {
                    throw error
                }

                lastError = error
            } catch let error as URLError where error.code == .timedOut {
                log("\(method) \(path) -> timeout (\(LogHelper.formatDuration(startTime)))")
                throw RenamedError.timeout(message: "Request timed out")
            } catch let error as URLError {
                log("\(method) \(path) -> network error (\(LogHelper.formatDuration(startTime)))")
                throw RenamedError.network(message: error.localizedDescription)
            } catch let error as RenamedError {
                throw error
            } catch {
                lastError = error
            }

            attempts += 1
            if attempts <= maxRetries {
                // Exponential backoff: 200ms, 400ms, 800ms, etc.
                let delayMs = Int(pow(2.0, Double(attempts)) * 100)
                log("Retry attempt \(attempts)/\(maxRetries), waiting \(delayMs)ms")
                try await Task.sleep(nanoseconds: UInt64(delayMs) * 1_000_000)
            }
        }

        throw lastError ?? RenamedError.network(message: "Request failed")
    }

    private func parseErrorMessage(from data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return json["error"] as? String ?? json["message"] as? String
    }

    /// Extract path from a URL string for logging purposes
    /// Returns the path portion without the base URL
    private func extractPath(from path: String) -> String {
        if path.hasPrefix("http://") || path.hasPrefix("https://") {
            // Full URL - extract just the path component
            if let url = URL(string: path) {
                return url.path
            }
            return path
        }
        // Already a path
        return path.hasPrefix("/") ? path : "/\(path)"
    }
}
